### ISSUE TITLE
aws_key_pair generates error

### ISSUE THREAD

**Issue by nwhite-dw** [hashicorp/terraform#11515]
1 error(s) occurred:

* aws_key_pair.terraform: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.8.4
    Resource ID: aws_key_pair.terraform
    Mismatch reason: extra attributes: public_key
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"key_name":*terraform.ResourceAttrDiff{Old:"", New:"terraform-key", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "fingerprint":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"key_name":*terraform.ResourceAttrDiff{Old:"", New:"terraform-key", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "public_key":*terraform.ResourceAttrDiff{Old:"", New:"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC9vk3mz3mhOt+x3qOTJyOGWJsHyPcGt9WHpDCKZtG4yTOAmFp7JY7UOst6W++SdDss6VcOvXWdjx7In1H3N1fDvZjR9wAj2iOYIoNtpVj0FVw/lmtmRUL2WPz1ItTBh48t1Vf2u8vMBy3AumSJ1GjfjxfgkAN3q+R0ttKQlNKBDOGKHMDpNQDAUcQLSojE99XUWs/qnQHgfkCb0oJTkvhsX9bzoqX31W0ukLrR/bcxlB8knDEc5Lu9stlU39n4z6eIwVF5S5GB6PJzIy8bCJjJYX8TbuEjN/rHqi3Qm9oFncUUuDl1DnyktvwRlmbEfVJ0aAqNaRDlmee2CrN2L1ah terraform", NewComputed:false, NewRemoved:false, NewExtra:"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC9vk3mz3mhOt+x3qOTJyOGWJsHyPcGt9WHpDCKZtG4yTOAmFp7JY7UOst6W++SdDss6VcOvXWdjx7In1H3N1fDvZjR9wAj2iOYIoNtpVj0FVw/lmtmRUL2WPz1ItTBh48t1Vf2u8vMBy3AumSJ1GjfjxfgkAN3q+R0ttKQlNKBDOGKHMDpNQDAUcQLSojE99XUWs/qnQHgfkCb0oJTkvhsX9bzoqX31W0ukLrR/bcxlB8knDEc5Lu9stlU39n4z6eIwVF5S5GB6PJzIy8bCJjJYX8TbuEjN/rHqi3Qm9oFncUUuDl1DnyktvwRlmbEfVJ0aAqNaRDlmee2CrN2L1ah terraform\n", RequiresNew:true, Sensitive:false, Type:0x0}, "fingerprint":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_key_pair

### Expected Behavior
Terraform apply works and new key is created.

### Actual Behavior
I got the error I pasted above.

**Comment by nwhite-dw** [comment]
Not sure if this matters but the .pub key file is simply an empty file meant to satisfy ```public_key = "${file("vpc/ssh/terraform-deployer.pub")}"``` which expects to see the file there even though this resource depends on the local-exec to run first and generate it.

**Comment by armon** [comment]
Just ran into this exact issue. The second `terraform apply` works as the `null_resource` has already created the key pair.

**Comment by mitchellh** [comment]
This is tricky. The `file` isn't being marked as computed since it is checking the file at this moment. It needs to be deferred and treated as "computed" to make this work... I'm actually not sure what the correct way to handle this is atm.

**Comment by armon** [comment]
@mitchellh Could we have a `file` data source that is part of the graph? That way it's input can be a path and output can be computed, it would just act like a normal resource in the graph. I think it's the HIL ahead-of-time evaluation that is causing the issue.

**Comment by mitchellh** [comment]
Yeah it is, and thats exactly what we need except that isn't possible at the moment. We don't have any functions that return "computed". I'll have to think on it.

**Comment by apparentlymart** [comment]
@armon the `template_file` data source can be (ab)used for that in most cases, as long as the file to be loaded doesn't happen to contain things that look like HIL interpolation markers.

I think any time someone finds themselves trying to write a file to disk and re-read it later this suggests a more fundamental hole in Terraform's functionality. In this specific case the `tls_private_key` resource was augmented with support for generating OpenSSH-style public key strings for a generated private key without creating any temporary files on disk.
